File: trunk/CharasLike/Windows/ImagePacker.vb
Public Class ImagePacker

    Private Sub Form7_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
    End Sub

    Shared Function ContainsThisBox(ByVal Controls As Control.ControlCollection, ByVal Box As RPGPictureBox)
        If Controls Is Nothing Or Box Is Nothing Then Return False

        For Each Control In Controls
            Dim MyBox = CType(Control, RPGPictureBox)
            If MyBox.MyImage Is Box.MyImage Then Return True
        Next
        Return False
    End Function

    Private Sub TreeView1_NodeMouseClick(ByVal sender As System.Object, ByVal e As System.Windows.Forms.TreeNodeMouseClickEventArgs) Handles TreeView1.NodeMouseClick
        ' Clear current box
        FlowLayoutPanel1.Controls.Clear()

        Dim Node As GameNodeNode = CType(e.Node, GameNodeNode)

        If Node.RPGNode.Images.Count Then
            For Each Image In Node.RPGNode.Images
                ' Add a box
                If ContainsThisBox(FlowLayoutPanel2.Controls, Image.Box) = False Then
                    FlowLayoutPanel1.Controls.Add(Image.Box.CreatePictureBoxCopy())
                End If
            Next
        End If
        TreeView1.SelectedNode = Node
    End Sub

    Private Sub FlowLayoutPanel2_DragEnter(ByVal sender As System.Object, ByVal e As System.Windows.Forms.DragEventArgs) Handles FlowLayoutPanel2.DragEnter
        If e.Data.GetDataPresent(GetType(RPGPictureBox)) = True Then
            Dim Box As RPGPictureBox = e.Data.GetData(GetType(RPGPictureBox))

            If ContainsThisBox(FlowLayoutPanel2.Controls, Box) Then
                e.Effect = DragDropEffects.None
            Else
                e.Effect = DragDropEffects.Copy
            End If
        Else
            e.Effect = DragDropEffects.None
        End If
    End Sub

    Private Sub FlowLayoutPanel2_DragDrop(ByVal sender As System.Object, ByVal e As System.Windows.Forms.DragEventArgs) Handles FlowLayoutPanel2.DragDrop
        If e.Data.GetDataPresent(GetType(RPGPictureBox)) = True Then
            Dim Box As RPGPictureBox = e.Data.GetData(GetType(RPGPictureBox))

            FlowLayoutPanel2.Controls.Add(Box)
        End If
    End Sub

    Private Sub FlowLayoutPanel1_DragEnter(ByVal sender As System.Object, ByVal e As System.Windows.Forms.DragEventArgs) Handles FlowLayoutPanel1.DragEnter
        If e.Data.GetDataPresent(GetType(RPGPictureBox)) = True Then
            Dim Box As RPGPictureBox = e.Data.GetData(GetType(RPGPictureBox))

            If ContainsThisBox(FlowLayoutPanel1.Controls, Box) Then
                e.Effect = DragDropEffects.None
            Else
                e.Effect = DragDropEffects.Copy
            End If
        Else
            e.Effect = DragDropEffects.None
        End If
    End Sub

    Private Sub FlowLayoutPanel1_DragDrop(ByVal sender As System.Object, ByVal e As System.Windows.Forms.DragEventArgs) Handles FlowLayoutPanel1.DragDrop
        If e.Data.GetDataPresent(GetType(RPGPictureBox)) = True Then
            Dim Box As RPGPictureBox = e.Data.GetData(GetType(RPGPictureBox))
            FlowLayoutPanel1.Controls.Add(Box)
        End If
    End Sub

    Private Sub Button2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button2.Click
        Me.DialogResult = Windows.Forms.DialogResult.Cancel
        Me.Close()
    End Sub

    Private Sub Button1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button1.Click
        Me.DialogResult = Windows.Forms.DialogResult.OK
        Me.Close()
    End Sub

    Private Sub PackEverythingToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles PackEverythingToolStripMenuItem.Click
        FlowLayoutPanel2.Controls.Clear()

        For Each Node In Images.RPGNodes
            Node.RecurseNodeAndAddAll()
        Next

        FlowLayoutPanel1.Controls.Clear()
    End Sub

    Private Sub UnpackEverythingToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles UnpackEverythingToolStripMenuItem.Click
        FlowLayoutPanel2.Controls.Clear()
    End Sub

    Private Sub PackAllToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles PackAllToolStripMenuItem.Click
        While FlowLayoutPanel1.Controls.Count <> 0
            FlowLayoutPanel2.Controls.Add(FlowLayoutPanel1.Controls(0))
        End While
    End Sub

    Private Sub UnpackAllToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles UnpackAllToolStripMenuItem.Click
        FlowLayoutPanel2.Controls.Clear()
    End Sub
End Class

File: Windows/ImagePacker.cs
using System.Collections.Generic;
using System;
using System.Drawing;
using System.Diagnostics;
using System.Collections;
using System.Windows.Forms;

namespace OpenCharas
{
	public partial class ImagePacker
	{
		public ImagePacker()
		{
			InitializeComponent();
		}

		public void Form7_Load(System.Object sender, System.EventArgs e)
		{
		}

		static public bool ContainsThisBox(Control.ControlCollection Controls, RPGPictureBox Box)
		{
			if (Controls == null || Box == null)
				return false;

			foreach (var Control in Controls)
			{
				var MyBox = (RPGPictureBox)Control;
		
				if (MyBox.MyImage == Box.MyImage)
					return true;
			}

			return false;
		}

		public void TreeView1_NodeMouseClick(System.Object sender, System.Windows.Forms.TreeNodeMouseClickEventArgs e)
		{
			// Clear current box
			FlowLayoutPanel1.Controls.Clear();

			GameNodeNode Node = (GameNodeNode)e.Node;

			if (Node.RPGNode.Images.Count != 0)
			{
				foreach (var Image in Node.RPGNode.Images)
				{
					// Add a box
					if (ContainsThisBox(FlowLayoutPanel2.Controls, Image.Box) == false)
						FlowLayoutPanel1.Controls.Add(Image.Box.CreatePictureBoxCopy());
				}
			}

			TreeView1.SelectedNode = Node;
		}

		public void FlowLayoutPanel2_DragEnter(System.Object sender, System.Windows.Forms.DragEventArgs e)
		{
			if (e.Data.GetDataPresent(typeof(RPGPictureBox)))
			{
				RPGPictureBox Box = (RPGPictureBox)e.Data.GetData(typeof(RPGPictureBox));

				if (ContainsThisBox(FlowLayoutPanel2.Controls, Box))
					e.Effect = DragDropEffects.None;
				else
					e.Effect = DragDropEffects.Copy;
			}
			else
				e.Effect = DragDropEffects.None;
		}

		public void FlowLayoutPanel2_DragDrop(System.Object sender, System.Windows.Forms.DragEventArgs e)
		{
			if (e.Data.GetDataPresent(typeof(RPGPictureBox)))
				FlowLayoutPanel2.Controls.Add((RPGPictureBox)e.Data.GetData(typeof(RPGPictureBox)));
		}

		public void FlowLayoutPanel1_DragEnter(System.Object sender, System.Windows.Forms.DragEventArgs e)
		{
			if (e.Data.GetDataPresent(typeof(RPGPictureBox)))
			{
				RPGPictureBox Box = (RPGPictureBox)e.Data.GetData(typeof(RPGPictureBox));

				if (ContainsThisBox(FlowLayoutPanel1.Controls, Box))
					e.Effect = DragDropEffects.None;
				else
					e.Effect = DragDropEffects.Copy;
			}
			else
				e.Effect = DragDropEffects.None;
		}

		public void FlowLayoutPanel1_DragDrop(System.Object sender, System.Windows.Forms.DragEventArgs e)
		{
			if (e.Data.GetDataPresent(typeof(RPGPictureBox)))
				FlowLayoutPanel1.Controls.Add((RPGPictureBox)e.Data.GetData(typeof(RPGPictureBox)));
		}

		public void Button2_Click(System.Object sender, System.EventArgs e)
		{
			this.DialogResult = System.Windows.Forms.DialogResult.Cancel;
			this.Close();
		}

		public void Button1_Click(System.Object sender, System.EventArgs e)
		{
			this.DialogResult = System.Windows.Forms.DialogResult.OK;
			this.Close();
		}

		public void PackEverythingToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
		{
			FlowLayoutPanel2.Controls.Clear();

			foreach (var Node in Images.RPGNodes)
				Node.RecurseNodeAndAddAll();

			FlowLayoutPanel1.Controls.Clear();
		}

		public void UnpackEverythingToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
		{
			FlowLayoutPanel2.Controls.Clear();
		}

		public void PackAllToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
		{
			while (FlowLayoutPanel1.Controls.Count != 0)
				FlowLayoutPanel2.Controls.Add(FlowLayoutPanel1.Controls[0]);
		}

		public void UnpackAllToolStripMenuItem_Click(System.Object sender, System.EventArgs e)
		{
			FlowLayoutPanel2.Controls.Clear();
		}
	}
}
